### ISSUE TITLE
[Roadmap] Further Ngram Speculative Decoding Support

### ISSUE THREAD

**Issue by kpham-sgl** [sgl-project/sglang#21052]
## Overview

Ngram speculative decoding ([paper](https://arxiv.org/pdf/2312.12728)): during generation, previously decoded tokens are inserted into a trie. Before each forward pass, the trie is queried with the current suffix to produce a draft token tree, constructed via BFS (recency) or priority queue (frequency). The target model then verifies the entire tree in one pass. No draft model needed.

## Limitations

1. **No external corpus support.** The trie is only populated from the current decoding session's output tokens. Ngram speculative decoding works best with a large reference corpus, but there is currently no mechanism to load one.

2. **Insert path does not scale to long inputs.** It builds trie paths for almost every suffix, leading to near-O(n²) memory growth; when capacity is full, eviction can only remove leaf nodes and may fail.

## Goals

1. Support external corpus lookup
2. Scale to long input prefills

### Refactor

- [x] Split monolithic C++ ngram module into separate files — [PR #20393](https://github.com/sgl-project/sglang/pull/20393) 
- [x] Rename `branch_length` to `max_trie_depth`- [PR #21181](https://github.com/sgl-project/sglang/pull/21181)
- [x] Remove `max_match_window_size` and `min_match_window_size` to match all suffixes in the trie - [PR #21225](https://github.com/sgl-project/sglang/pull/21225)
- [x] Maintain per-anchor matching state across decode steps — advance anchors on each new token instead of re-walking from trie root - [PR #21243](https://github.com/sgl-project/sglang/pull/21243)
- [x] Fix race condition in `TrieCache::insert()` when the worker thread's queue is empty - [PR #21186](https://github.com/sgl-project/sglang/pull/21186)
- [x] Replace busy-wait polling in `Ngram<Cache>::synchronize()` with a condition variable - [PR #21186](https://github.com/sgl-project/sglang/pull/21186)

### New Features

- [x] Load an external corpus (pre-specified format) via a server arg and build a [Suffix Automaton (SAM)](https://cp-algorithms.com/string/suffix-automaton.html) over it - [PR #21425](https://github.com/sgl-project/sglang/pull/21425)
- [x] Support multiple SAMs with dynamic add/remove via HTTP API - [PR #22203](https://github.com/sgl-project/sglang/pull/22203)
- [ ] Enhance support for mult-SAMs - [PR #22294](https://github.com/sgl-project/sglang/pull/22294)
- [x] Accept length benchmark: output-as-corpus proves SAM ≥2x accept length boost - [PR #22199](https://github.com/sgl-project/sglang/pull/22199)
- [ ] Better solution for `combineResults`. Instead of relying on fixed `external_sam_budget`, use weighted dynamic allocation for Trie, SAM(s) match tokens based on longer matching / recency / frequency. 
- [ ] SAMs eviction. Evict by least recently used SAM to make space
- [ ] Hierarchical SAM / Trie construction and matching system (per-request, per-corpus-id, global)

**Comment by satyamk7054** [comment]
@kpham-sgl Thank you for the detailed roadmap. I wanted to check on the below 2 items (it seems they are not planned in the roadmap but helped my use-case – I can help with contribution if required):

1. Prompt Seeding for Ngram
Currently we [skip injecting the prompt text](https://github.com/sgl-project/sglang/blob/c89afaea7cbdd25e29ebcaf1511bf32b398cfb8d/python/sglang/srt/speculative/ngram_worker.py#L237) into trie but for my use-case, prompt seeding helped improve latency by ~10%. I think there'd be concerns on defaulting this to "enabled" without perf improvement for much larger texts (my use-case had ~7k avg prompt tokens). 
Are any improvements planned for prompt seeding / ingestion path (probably same mechanics as external corpus) or we can put this behind a server-arg for now

2. SGLang seems to [restrict concurrency to 48 requests for ngram spec-decode](https://github.com/sgl-project/sglang/blob/c89afaea7cbdd25e29ebcaf1511bf32b398cfb8d/python/sglang/srt/server_args.py#L3278). I think that's not required for ngram (might be a carryover from other spec-decode code). Are there any concerns to increasing / removing this check? It affects TTFT especially when the ngram trie has not warmed up (prompt seeding also helps warmup)

Additionally there may be some cross-request pollution for my use-case specifically that results in slightly lower acceptance rate but I think request-level trie is already in scope in the above plan

**Comment by kpham-sgl** [comment]
1. I should be a bit more clear in my roadmap. The last item (along with extending SAM match length) cover this use case. We intend to create a per-request SAM (constructed from a request prompt). 
Is this similar to what you meant by Prompt Seeding? 
2. I don't have immediate answer here and will take a deeper look

Re: cross-request pollution, thanks for raising awareness here. We intend to make request-level trie for security reasons (side-channel attack et al) but this is a good motivation also

**Comment by bandham-manikanta** [comment]
Hi @kpham-sgl — on the ngram default `max_running_requests=48` when unset: is investigating that still open for an outside contributor, or already covered in the ngram series?

If open, I can do a small A100 A/B (48 vs higher concurrency) and either report here or open a narrow PR. I won’t start a PR until you confirm the scope is free.

**Comment by shiloz-stack** [comment]
Hi @kpham-sgl — I noticed the [TODO](https://github.com/sgl-project/sglang/blob/99b29bf1889c7bf84973596a5273adc80a15b523/python/sglang/srt/speculative/ngram_worker.py#L300) in `ngram_worker.py` about replacing the per-request tree-mask construction loop with a kernel. I’m interested in picking this up. My initial plan is to:
1. Add a correctness test comparing the current implementation with the kernel output across mixed sequence lengths and batch sizes.
2. Implement a Triton kernel for constructing the flattened full tree mask.
3. Run a single-GPU microbenchmark across increasing batch sizes and report the results.

Is anyone already working on this optimization? Let me know if I can work on this, happy to help!

**Comment by bandham-manikanta** [comment]
Following up on my earlier note about the ngram default `max_running_requests=48`:

> Hi @kpham-sgl — on the ngram default `max_running_requests=48` when unset: is investigating that still open for an outside contributor, or already covered in the ngram series?
>
> If open, I can do a small A100 A/B (48 vs higher concurrency) and either report here or open a narrow PR. I won’t start a PR until you confirm the scope is free.

That item still looks free (separate from the tree-mask Triton work in #32407). I’m going to start the A100 A/B and open a narrow draft PR if the numbers support raising/removing the cap for ngram. Will link it here. Please shout if this is already claimed or you’d rather keep the limit as-is for now.

**Comment by bandham-manikanta** [comment]
Following up on my earlier note about starting the A100 A/B for the ngram default `max_running_requests=48`:

> Following up on my earlier note about the ngram default `max_running_requests=48`:
>
> > Hi @kpham-sgl — on the ngram default `max_running_requests=48` when unset: is investigating that still open for an outside contributor, or already covered in the ngram series?
> >
> > If open, I can do a small A100 A/B (48 vs higher concurrency) and either report here or open a narrow PR. I won’t start a PR until you confirm the scope is free.
>
> That item still looks free (separate from the tree-mask Triton work in #32407). I’m going to start the A100 A/B and open a narrow draft PR if the numbers support raising/removing the cap for ngram. Will link it here. Please shout if this is already claimed or you’d rather keep the limit as-is for now.

### Investigation update

I ran A100 A/B (NGRAM, Qwen3-8B) comparing forced default `max_running_requests=48` vs explicit `--max-running-requests 96` under overload. Higher concurrency improved TTFT and throughput; TPOT worsened as expected. So the forced `48` is a real concurrency bottleneck under load.

However, simply removing the forced `48` and leaving `max_running_requests=None` is **not a safe standalone fix** on my 1×A100 setup:

- With the force removed, default launch fails during CUDA graph capture (FlashInfer workspace OOM while capturing large decode/verify batch sizes, observed up to bs=256).
- The server only starts again if I also pass an explicit graph limit, e.g. `--cuda-graph-max-bs-decode 16`.
- So today, forced `max_running_requests=48` appears to serve **two roles**:
  1. concurrency default
  2. implicit CUDA-graph capture safety cap

**Conclusion:** request-concurrency defaulting and CUDA-graph capture batch-size safety should be separated. Auto path with `max_running_requests=None` does not currently launch safely by itself for NGRAM on this setup.

Happy to open/adjust a draft PR once maintainers prefer a direction, e.g.:
- remove the forced `48` **and** set/cap decode CUDA graph max BS when unset for NGRAM, or
- keep a temporary concurrency default until graph sizing is fixed independently.

@kpham-sgl would appreciate guidance on preferred approach.

**Comment by bandham-manikanta** [comment]
Confirmed with @Khoa-Pham: max_running_requests and CUDA graph max decode batch size are independent flags in SGLang, but should be aligned in deployment depending on VRAM capacity to avoid FlashInfer workspace OOMs during graph capture.

Thanks.

**Comment by JasonKeyiL** [comment]
Hi @kpham-sgl @Qiaolin-Yu, can I take "Expand SAM match length (currently still cap by max_trie_depth)"?

Both `buildRecency` and `buildFrequency` call `match(context, len, param.max_trie_depth)` (suffix_automaton.cpp:183 and :219), and `match()` cuts the context down to its last `max_depth` tokens. The trie needs that limit because it only stores paths that deep, but the SAM holds the whole corpus and can match a much longer suffix for free, so we are throwing away the main reason to have it.

I put a small standalone harness on top of suffix_automaton.cpp to see what it costs. 64 passages, each one 8 unique tokens + 12 shared boilerplate + 8 unique continuation. The query is a passage prefix plus the boilerplate, so only context older than the boilerplate tells you which passage to continue. Measuring how much of the true continuation is reachable in the draft tree (draft_token_num=8, bfs_breadth 1..4):

```
match_depth   mean accept len   fully drafted
     4            0.125             1/64
    12            0.125             1/64
    16            8.000            64/64
```

The cliff is exactly at the boilerplate length. With the default max_trie_depth=18 you need more than 18 tokens of shared context before it bites, which does not look far-fetched for a RAG-style corpus.

One question before I write it. max_trie_depth is also the scale for the BFS breadth right under the match call:

```cpp
const auto max_match_depth = std::max<int32_t>(1, int32_t(param.max_trie_depth - 1));
const double bfs_breadth_scale = double(param.max_bfs_breadth - param.min_bfs_breadth) / max_match_depth;
```

so once matched_len can exceed max_trie_depth the starting breadth goes negative and clamps to 1, and a long match degenerates into a single chain. My plan is a separate `max_sam_match_depth` for the SAM match, clamp matched_len when computing the breadth so the existing curve stays calibrated, and leave trie.cpp alone since its bound is structural. Does that sound right, or did you want the SAM to just match unbounded?

**Comment by JasonKeyiL** [comment]
Correcting myself: the C++ cap is not the only one, and on its own it is not even the binding one.

`ngram_worker.py` already trims the query before it ever reaches the SAM. The tails are sliced at lines 260 and 263, then `_efficient_concat_last_n(..., self.max_trie_depth)` at 296 trims again, and there is a comment at 257 spelling out the assumption ("Only the last max_trie_depth tokens can match"). So the SAM never sees more than `max_trie_depth` tokens, and changing `suffix_automaton.cpp` alone would be a no-op in production.

That makes this bigger than I first said: the worker has to carry a longer tail too, which costs a little extra per request on every decode step, so the default becomes a real decision rather than a detail.

My plan now is to add `speculative_ngram_max_sam_match_depth` defaulting to 0, meaning "same as today, SAM capped at max_trie_depth", so the longer match is opt-in and nothing regresses. We can argue about a better default separately once there are numbers on a real corpus. Say the word if you'd rather have it on by default.

One thing I still need to check before touching it: `Trie::buildRecency` clamps with `std::min(len, param_.max_trie_depth)` itself so a longer context looks harmless there, but the stateful path keeps per-anchor state sized against `max_trie_depth`, and I want to be sure a longer tail does not confuse the incremental anchor advance.

**Comment by chengcuiping** [comment]
Hi @kpham-sgl, I’d like to work on the unchecked **“SAM for user request’s input”** item.

My current understanding is to build an ephemeral, request-scoped SAM from each request’s prompt token IDs and use it as an additional NGRAM proposal source during decoding. The prompt SAM should not be registered as a global external corpus or become visible to other requests.

For the first PR, I propose the following scope:

* Build and query a prompt SAM per request, with bounded construction cost and memory usage.
* Integrate its candidates with the existing per-request Trie and external-SAM proposal path.
* Preserve the current behavior when prompt SAM is not enabled.
* Define and test the lifecycle for request completion, abort, and retraction/rescheduling, ensuring there is no cross-request prompt leakage.
* Add correctness tests for concurrent requests.
* Benchmark SAM construction overhead, CPU memory usage, acceptance length, TTFT, TPOT, and end-to-end throughput on representative coding/agentic prompts.

Before I begin implementation, could you please confirm:

1. Is this the intended semantics of this roadmap item?
2. Should this work build on or interoperate with #35102 and #37879, or should the first PR remain independent?
3. Is anyone already working on this item, and is there a preferred maintainer/reviewer with whom I should coordinate?

If this scope looks reasonable, I can first post a short design note and then implement a minimal single-scheduler version. I would leave request-level `corpus_ids` selection and multi-scheduler SAM transport to their separate roadmap items.

Thanks!

**Comment by hkr04** [comment]
Hi @kpham-sgl ,  this may also be relevant to the multi-source NGRAM direction here: I recently opened #39211 implementing **RACER**, which combines history retrieval with target-logit reuse for training-free tree drafting.

The idea is to reuse top-k alternatives already produced during target verification to fill draft slots that history retrieval cannot cover. On Qwen3.8-27B/A800, this improves K32 acceptance from 1.89 to 2.50 and throughput from 37.7 to 45.3 tok/s over same-K NGRAM.

I also found that a simpler **NGRAM + `min_match_depth` (or `min_match_window_size`) + logits-assisted expansion** behaves similarly, so I plan to explore that as a separate, smaller NGRAM extension. Leaving this here since it seems closely related to the `combineResults` / multi-source candidate allocation direction in this roadmap.
